### PR TITLE
Re-factor ceffu

### PR DESCRIPTION
**Pull request by mxiao-cll**
## Description

......

## Changes

- High level
- changes that
- you made

## Steps to Test

1. Steps
2. to
3. test

## Quality Assurance

- [ ] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `infra-k8s` configuration file.
- [ ] If a new adapter was made, or an existing one was modified so that its environment variables have changed, update the relevant `adapter-secrets` configuration file or update the [soak testing blacklist](/packages/scripts/src/get-changed-adapters/soakTestBlacklist.ts).
- [ ] If a new adapter was made, or a new endpoint was added, update the `test-payload.json` file with relevant requests.
- [ ] The branch naming follows git flow (`feature/x`, `chore/x`, `release/x`, `hotfix/x`, `fix/x`) or is created from Jira.
- [ ] This is related to a maximum of one Jira story or GitHub issue.
- [ ] Types are safe (avoid TypeScript/TSLint features like any and disable, instead use more specific types).
- [ ] All code changes have 100% unit and integration test coverage. If testing is not applicable or too difficult to justify doing, the reasoning should be documented explicitly in the PR.

### CODE REVIEW

File: .changeset/funny-lemons-invent.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@chainlink/ceffu-adapter': patch
+---
+
+Code refactor
```
Comment (dskloetc) [inline .changeset/funny-lemons-invent.md:5]: Changing retry behavior and exposing extra data in the response are not "refactoring".

File: packages/sources/ceffu/src/transport/mirrorX.ts [hunk 2]
```diff
@@ -1,28 +1,14 @@
...
 import { AdapterError } from '@chainlink/external-adapter-framework/validation/error'
 import { Decimal } from 'decimal.js'
-import { HttpsProxyAgent } from 'https-proxy-agent'
-import { stringify } from 'querystring'
-// TODO: Refactor to use request instead
-// https://smartcontract-it.atlassian.net/browse/OPDATA-4224
-import { sign } from './wallet/requester'
+import { request } from './requester'
 
 interface ApiResponse {
```
Comment (dskloetc) [inline packages/sources/ceffu/src/transport/mirrorX.ts]: Is `ApiResponse` still the right name? We don't control the API so refactoring the code can't really change the type of the response we get from the API.

File: packages/sources/ceffu/src/transport/mirrorX.ts [hunk 3]
```diff
@@ -35,50 +21,33 @@ export const getAssetPositions = async (
...
+        privateKey,
+        requester,
+        proxy,
+      )
     }),
   )
 
   for (const response of responses) {
-    if (!response.response || !response.response.data) {
+    if (!response || !response.extra || response.extra.length == 0) {
```
Comment (dskloetc) [inline packages/sources/ceffu/src/transport/mirrorX.ts]: Prefer `===` over `==`. For example `false == 0` is true which can be confusing.
Or optionally shorten this to
```
if (!response?.extra?.length) {
```

File: packages/sources/ceffu/src/transport/mirrorX.ts [hunk 4]
```diff
@@ -35,50 +21,33 @@ export const getAssetPositions = async (
...
-
-    if (!response.response.data.data?.exchangeBalance) {
-      throw new AdapterError({
-        statusCode: 500,
-        message: `Ceffu API does not return exchangeBalance, code: ${response.response.data.code}, message:${response.response.data.message}`,
-      })
-    }
   }
 
+  const exchangeBalances = responses.flatMap(({ extra }) =>
```
Comment (dskloetc) [inline packages/sources/ceffu/src/transport/mirrorX.ts:45]: What is "extra". Why do we need to access `extra` here while previously we could just use `response.data.data.exchangeBalance`?

File: packages/sources/ceffu/src/transport/requester.ts
Comment (dskloetc) [inline packages/sources/ceffu/src/transport/requester.ts:1]: Moving and changing the file in the same commit makes it hard to review.
Can you keep the file in the same place for now, or separate out the changes to this file?

File: packages/sources/ceffu/src/transport/wallet/wallet.ts [hunk 5]
```diff
@@ -26,14 +26,14 @@ export const getWallets = async (
     proxy,
   )
 
-  if (!response || response.length == 0) {
+  if (!response || !response.data || response.data.length == 0) {
```
Comment (dskloetc) [inline packages/sources/ceffu/src/transport/wallet/wallet.ts]: Use `===` or replace `response.data.length == 0` with `!response.data.length` depending on intent.

File: .changeset/funny-lemons-invent.md (hunk 1 above)
Comment (mxiao-cll) [reply .changeset/funny-lemons-invent.md:5]: Removed logic change

File: packages/sources/ceffu/src/transport/mirrorX.ts (hunk 2 above)
Comment (mxiao-cll) [reply packages/sources/ceffu/src/transport/mirrorX.ts]: renamed

File: packages/sources/ceffu/src/transport/mirrorX.ts (hunk 4 above)
Comment (mxiao-cll) [reply packages/sources/ceffu/src/transport/mirrorX.ts:45]: Because the exchangeBalance is now packed into the extra array, we can not fit this into the data array due to exchangeBalance sit outside of the data object in response.

File: packages/sources/ceffu/src/transport/requester.ts
Comment (mxiao-cll) [reply packages/sources/ceffu/src/transport/requester.ts:1]: Moved back

File: packages/sources/ceffu/test/unit/requester.test.ts [hunk 6]
```diff
@@ -67,7 +67,7 @@ describe('requester.ts', () => {
 
       const result = await request('', '', {}, '', '', mockRequester)
 
-      expect(result).toEqual([100, 200])
+      expect(result).toEqual({ data: [100, 200], extra: [] })
```
Comment (dskloetc) [inline packages/sources/ceffu/test/unit/requester.test.ts:70]: Can you add a unit test with non-empty `extra`?

File: packages/sources/ceffu/src/transport/rate.ts [hunk 7]
```diff
@@ -28,6 +28,7 @@ export const toUsd = async (
         rate: rate.value,
         decimal: rate.decimal,
         value,
+        data: c.data,
```
Comment (dskloetc) [inline packages/sources/ceffu/src/transport/rate.ts]: How is this extra field used?
Is it related to the refactoring or new functionality?

File: packages/sources/ceffu/src/transport/wallet/requester.ts [hunk 8]
```diff
@@ -19,19 +24,29 @@ type EmptyParams = Record<string, never>
...
 }
+type MirrorXParams = {
+  mirrorXLinkId: string
+  excludeZeroAmountFlag: boolean
+}
 
 const successCode = '000000'
 
-export const request = async <T>(
+export const request = async <T, E = EmptyParams>(
```
Comment (dskloetc) [inline packages/sources/ceffu/src/transport/wallet/requester.ts]: Can you add some documenting comments explaining that `data` will have 1 array with all items across pages but that `extra` will only have 1 item per page?

File: packages/sources/ceffu/src/transport/rate.ts (hunk 7 above)
Comment (dskloetc) [inline packages/sources/ceffu/src/transport/rate.ts]: If the goal is just to augment the `coins` array, would it make sense to do
```
      return {
        ...c,
        coin: c.coin.toUpperCase(),
        rate: rate.value,
        decimal: rate.decimal,
        value,
      }
```
instead? (add `...c,` and remove `amount: c.amount` and `data: c.data`)

Comment (mxiao-cll) [reply packages/sources/ceffu/src/transport/rate.ts]: It is currently not being used, we may use it later to funnel extra data through
